### DISCUSSION TITLE
State Management with Templ Web Framework

### DISCUSSION THREAD

**Discussion by CordialSocial** [a-h/templ#949] [Ideas]
GO/ Golang of itself, can handle backend effectively.

And when people these days want to build websites, they don't want to build something that is just simple-- else they will use something like **WordPress**, for a simple website.

The point is, if people want to commit time, hours and effort in coding a website from scratch, they actually want to build something **serious**, and with a lot of **interactivity**.

I mean, tools like **React**, **Vue**, and **Svelte** make it possible to create **Reusable Web Components**.

Their additional strong point is **State Management**-- where an application's components behaviour changes according to certain actions.

Can **Templ**  be upgraded to such height, as to provide the full features of tools like **React**, **Vue** and **Svelte**-- where we can create **Reusable Components**, as well as **State Management**-- in an easy to implement and easy to understand way?

Once **Templ** can be used to create **Reusable Web Components**, and Templ can handle **State Management** on the frontend, as well as cause these to be done in an easy to understand and easy to implement way, we would have brought the **React**, **Vue** and **Svelte** world to GO.

It will be nice if Templ can bring the full power of frontend -- which are **Reusable Web Components** and **State Management** capabilities to the world of GO.

The very things ( **Reusable Web Components** and **State Management** ) that make **React**, **Vue**, **Angular** and **Svelte** to be worth any mention.

Bring these same frontend capabilities to the GO Ecosystem.

**A POINT OF NOTE**
GO inherently can handle concurrency.

Use the concurrency feature of GO to manage **State** on a Web App developed with Templ.

What's the way forward on this?

Regards.

**Comment by a-h** [comment]
Not really sure what you're asking here.

The documentation is at https://templ.guide. It covers composition https://templ.guide/syntax-and-usage/template-composition/

There are component libraries being developed that target templ, e.g. https://goship.it/ - we're going to be adding them to https://github.com/templ-go/awesome-templ

I don't know what your specific question is related to state management. There's a sample of state management included at https://templ.guide/server-side-rendering/example-counter-application/

templ works well with HTMX and https://data-star.dev/ for adding client side interactivity, including the use of SSE for realtime updates to UI.

**Reply by CordSocial** [reply]
@a-h 

Your reply is well appreciated.  Your reply seems to open my eyes to what is available in **GO** and **Templ**.

Now, what I'm trying to ask for, is a fully featured **Frontend Web Framework** for GO.

The thing is, it gets frustrating learning and learning too much, just to achieve almost straight forward things like **State Management** and **Web Components**.

I just want to learn GO and develop a backend in Go.

Then learn one **Web Framework**-- maybe **TEMPL**, and that should be all I ever need to build a website like Facebook.

What this suggests, is that Templ should be capable of being a **Frontend Monolith** like **Svelte**, yet be able to be modular: also like Svelte.

The case of Svelte is that it has everything you need for frontend (Components, Animations, DOM Manipulation, Scoped CSS, JavaScript_{Logic}_ ) -- in which case you don't need any other external libraries or packages to build your user Interface and be productive in Svelte, and get frontend Interactivity in its most advanced form.

This means that **Templ** should have everything we ever need to build **Web Applications** in GO. It should be sufficient to handle all things frontend in GO, and it should have pluggable modules-- where we only install the modules we need for our project.

I don't want to go through the stress of learning Go, learning Templ, learning HTMx, learning DataStar, learn ORM, learn this and learn that, and learning whatever else, all because I want to build a website like Facebook.

GO and one Web Framework ( _maybe TEMPL_ ) should be enough: that's what I'm trying to say.

And that framework should of itself, be able to do whatever functionalities and features that **React**, **Vue**, or **Svelte**, brings to frontend Development.

**AN EXAMPLE USE CASE**
If I learn **Dart**, I can learn **ServerPod** and **Flutter**, and that's all I ever need to build a Mobile Application.

**Jasper** ( https://pub.dev/packages/jaspr)  is in the pipeline.

If I want to build a Website like Facebook in the Dart/Flutter world, I will learn Dart, ServerPod and **Jaspr**, and that's all I need to build a website in the Dart/ Flutter world, without needing anything else.

**IN SUMMARY**
Templ is defined as "**A language for writing HTML user interfaces in Go**".

**THE REQUEST**
TEMPL should upgrade and be more than just "**_A language for writing HTML user interfaces in Go_**".

It should upgrade to be "**_a fully featured Frontend Web Framework for GO_**"-- in which case, it will have all the full power of React, Vue, or Svelte.

By the full power of React, Vue, or Svelte, what I refer to here, is:

1.) Reusable and Reactive Web Components.
2.) State Management ( out of the box ).
3.) Styling.
4.) Animations and Transitions.
5.) and every beautiful frontend feature that Svelte has.

This is what I'm saying, and I do hope you now understand ?

**Reply by amoghyermalkar123** [reply]
If i understand what you are saying in short is, "I don't want to learn go/ htmx/ datastar, etc" Just plain and simple Templ? All you need is to build websites using web frameworks without learning the things that comprise them.

If this is what you mean then let me break some things down for you -
- You want to build websites that don't look like wordpress themed? You will have to learn concepts in depth, there is no two ways about it. You either don't want to put efforts to learn things fundamentally (so you use wordpress) or you want to build websites that serve a complex purpose.
- It's better to present the use-case when asking for things like this all you've defined under use-case section is i dont want to put efforts learning new things and understanding stuff fundamentally which to me says you are a junior dev/ just a designer.


Don't get me wrong, but your ask is vague and does not come as either a feedback/ valid use case.

**Comment by Riu** [comment]
@CordialSocial Templ is responsible for the HTML presentation layer. That's it. 

BTW: you are a software developer, have you coded anything with Templ? What you are asking and how it is asked seems strange, to say the least.
